File: tests/typesafety/python/exercise_lookup_solution.py
# expect: passes
# run: yes
# title: Each lookup returns a Result with its own reason
# improves: exercise_lookup_start
"""The same three steps, with a Result. Each failure carries its reason, and
`and_then` threads the chain."""

from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True)
class Ok[T]:
    value: T


@dataclass(frozen=True)
class Err:
    reason: str


type Result[T] = Ok[T] | Err


def and_then[T, U](result: Result[T], step: Callable[[T], Result[U]]) -> Result[U]:
    match result:
        case Ok(value):
            return step(value)
        case Err():
            return result


def find_user_id(login: str) -> Result[str]:
    return Ok("u7") if login == "ada" else Err(f"no user for login {login!r}")


def find_cart(user_id: str) -> Result[list[str]]:
    return Ok(["book"]) if user_id == "u7" else Err(f"no cart for {user_id}")


def head(items: list[str]) -> Result[str]:
    return Ok(items[0]) if items else Err("the cart is empty")


def first_item(login: str) -> Result[str]:
    return and_then(and_then(find_user_id(login), find_cart), head)


assert first_item("ada") == Ok("book")
assert first_item("bob") == Err("no user for login 'bob'")

File: tests/typesafety/typescript/exercise_lookup_solution.ts
// expect: passes

type Ok<T> = { readonly kind: "ok"; readonly value: T };
type Err = { readonly kind: "err"; readonly reason: string };
type Result<T> = Ok<T> | Err;

function ok<T>(value: T): Result<T> {
  return { kind: "ok", value };
}

function err(reason: string): Result<never> {
  return { kind: "err", reason };
}

function andThen<T, U>(result: Result<T>, step: (value: T) => Result<U>): Result<U> {
  return result.kind === "ok" ? step(result.value) : result;
}

function findUserId(login: string): Result<string> {
  return login === "ada" ? ok("u7") : err(`no user for login ${login}`);
}

function findCart(userId: string): Result<string[]> {
  return userId === "u7" ? ok(["book"]) : err(`no cart for ${userId}`);
}

function head(items: string[]): Result<string> {
  const first = items[0];
  return first === undefined ? err("the cart is empty") : ok(first);
}

export function firstItem(login: string): Result<string> {
  return andThen(andThen(findUserId(login), findCart), head);
}
